[SPARK-14335][SQL] Describe function command returns wrong output

yongtang opened the pull request:
## What changes were proposed in this pull request?

…because some of built-in functions are not in function registry.

This fix tries to fix issues in `describe function` command where some of the outputs
still shows Hive's function because some built-in functions are not in FunctionRegistry.

The following built-in functions have been added to FunctionRegistry:

```
-
!
*
/
&
%
^
+
<
<=
<=>
=
==
>
>=
|
~
and
in
like
not
or
rlike
when
```

The following listed functions are not added, but hard coded in `commands.scala` (@hvanhovell):

```
!=
<>
between
case
```

Below are the existing result of the above functions that have not been added:

```
spark-sql> describe function `!=`;
Function: <>
Class: org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPNotEqual
Usage: a <> b - Returns TRUE if a is not equal to b
```

```
spark-sql> describe function `<>`;
Function: <>
Class: org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPNotEqual
Usage: a <> b - Returns TRUE if a is not equal to b
```

```
spark-sql> describe function `between`;
Function: between
Class: org.apache.hadoop.hive.ql.udf.generic.GenericUDFBetween
Usage: between a [NOT] BETWEEN b AND c - evaluate if a is [not] in between b and c
```

```
spark-sql> describe function `case`;
Function: case
Class: org.apache.hadoop.hive.ql.udf.generic.GenericUDFCase
Usage: CASE a WHEN b THEN c [WHEN d THEN e]* [ELSE f] END - When a = b, returns c; when a = d, return e; else return f
```
## How was this patch tested?

Existing tests passed. Additional test cases added.

yongtang committed:
[SPARK-14335][SQL] Describe function command returns wrong output because some of built-in functions are not in function registry.

This fix tries to fix issues in `describe function` command where some of the outputs
still shows Hive's function because some built-in functions are not in FunctionRegistry.

The following built-in functions have been added to FunctionRegistry:
```
-
!
*
/
&
%
^
+
<
<=
<=>
=
==
>
>=
|
~
and
in
like
not
or
rlike
when
```

The following listed functions are not added:
```
!=
<>
between
case
```

All existing tests passed.

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #2735 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2735/consoleFull)** for PR 12128 at commit [`7a84e9b`](https://github.com/apache/spark/commit/7a84e9bd7dccffce103862a7c8818396d6fecb34).

SparkQA commented:
**[Test build #2735 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2735/consoleFull)** for PR 12128 at commit [`7a84e9b`](https://github.com/apache/spark/commit/7a84e9bd7dccffce103862a7c8818396d6fecb34).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

hvanhovell commented:
@yhuai what do you want to do with the composed expressions, i.e.: `!=`, `<>` & `between`? `case` is more of a keyword.

yhuai commented:
For composed expressions, if there are only a few of them, maybe we can have special handling in the describe command? For `case`, we do not need to support `describe function case` if it is more of a keyword.

yongtang commented:
Hi @yhuai @hvanhovell , as a reference, below is the result of the `describe function` for current master. I haven't been able to identify the corresponding functions:

```
spark-sql> describe function `!=`;
Function: <>
Class: org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPNotEqual
Usage: a <> b - Returns TRUE if a is not equal to b
```
```
spark-sql> describe function `<>`;
Function: <>
Class: org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPNotEqual
Usage: a <> b - Returns TRUE if a is not equal to b
```
```
spark-sql> describe function `between`;
Function: between
Class: org.apache.hadoop.hive.ql.udf.generic.GenericUDFBetween
Usage: between a [NOT] BETWEEN b AND c - evaluate if a is [not] in between b and c
```
```
spark-sql> describe function `case`;
Function: case
Class: org.apache.hadoop.hive.ql.udf.generic.GenericUDFCase
Usage: CASE a WHEN b THEN c [WHEN d THEN e]* [ELSE f] END - When a = b, returns c; when a = d, return e; else return f
```

hvanhovell commented:
@yongtang can you see if you can hardcode the descriptions for the composed function into the `DecribeFunction` command? See: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala#L421-L423

yongtang commented:
@hvanhovell Thanks. Let me take a look and update the pull request.

yongtang committed:
[SPARK-14335][SQL] Describe function command returns wrong output because some of built-in functions are not in function registry.

Hard code describe functions for `<>`, `!=`, `between`, and `case` for now,
based on the feedback from @hvanhovell.

yongtang commented:
Hi @hvanhovell the pull request was updated. Let me know if there are any issues.

yhuai commented:
ok to test

SparkQA commented:
**[Test build #54829 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54829/consoleFull)** for PR 12128 at commit [`baad78f`](https://github.com/apache/spark/commit/baad78ff31815b8d9e77284713c4bb79d8f2a0f3).

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala:
```diff
@@ -438,20 +438,27 @@ case class DescribeFunction(
+    functionName match {
+      case "<>" | "!=" | "between" | "case" =>
+        Row(s"Function: $functionName") ::
+        Row(s"Usage: To be added.") :: Nil
```
Can we add the usage at here?

yongtang replied:
@yhuai Sure let me update the pull request shortly.

yongtang committed:
[SPARK-14335][SQL] Describe function command returns wrong output because some of built-in functions are not in function registry.

Add usages for `<>`, `!=`, `between`, and `case`, based on the feedback from @yhuai.

SparkQA commented:
**[Test build #54834 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54834/consoleFull)** for PR 12128 at commit [`f318e8c`](https://github.com/apache/spark/commit/f318e8c3208c3bf5aed6d2202fb71052410aab81).

yongtang commented:
Hi @yhuai the pull request has been updated with the usages being added. Let me know if there are any other issues. And thanks for the review.

SparkQA commented:
**[Test build #54829 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54829/consoleFull)** for PR 12128 at commit [`baad78f`](https://github.com/apache/spark/commit/baad78ff31815b8d9e77284713c4bb79d8f2a0f3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54829/
Test FAILed.

SparkQA commented:
**[Test build #54834 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54834/consoleFull)** for PR 12128 at commit [`f318e8c`](https://github.com/apache/spark/commit/f318e8c3208c3bf5aed6d2202fb71052410aab81).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54834/
Test FAILed.

yongtang committed:
[SPARK-14335][SQL] Describe function command returns wrong output because some of built-in functions are not in function registry.

Update SQLQuerySuite to reflect changes in `describe function`.

SparkQA commented:
**[Test build #54855 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54855/consoleFull)** for PR 12128 at commit [`6e1e879`](https://github.com/apache/spark/commit/6e1e87964c83c5b313f7dc7b910fc8f1849c08e1).

SparkQA commented:
**[Test build #54855 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54855/consoleFull)** for PR 12128 at commit [`6e1e879`](https://github.com/apache/spark/commit/6e1e87964c83c5b313f7dc7b910fc8f1849c08e1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54855/
Test FAILed.

SparkQA commented:
**[Test build #2743 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2743/consoleFull)** for PR 12128 at commit [`6e1e879`](https://github.com/apache/spark/commit/6e1e87964c83c5b313f7dc7b910fc8f1849c08e1).

SparkQA commented:
**[Test build #2743 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2743/consoleFull)** for PR 12128 at commit [`6e1e879`](https://github.com/apache/spark/commit/6e1e87964c83c5b313f7dc7b910fc8f1849c08e1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala:
```diff
@@ -438,20 +438,38 @@ case class DescribeFunction(
+        Row(s"Function: $functionName") ::
+        Row(s"Usage: between a [NOT] BETWEEN b AND c - " +
+          s"evaluate if a is [not] in between b and c") :: Nil
+      case "case" =>
```
Looks like we also need to match upper case letters?

yongtang replied:
Thanks @yhuai let me update the pull request.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala:
```diff
@@ -438,20 +438,38 @@ case class DescribeFunction(
+        Row(s"Usage: a != b - Returns TRUE if a is not equal to b") :: Nil
+      case "between" =>
+        Row(s"Function: $functionName") ::
+        Row(s"Usage: between a [NOT] BETWEEN b AND c - " +
```
`Usage: between a ...` => `Usage: a ...`?

yongtang replied:
@yhuai Yes that makes sense. The first `between` appears in the original output (with `org.apache.hadoop.hive.ql.udf.generic.GenericUDFBetween`) which seems to be an bug.

yongtang committed:
[SPARK-14335][SQL] Describe function command returns wrong output because some of built-in functions are not in function registry.

Take into consideration the upper and lower case of `describe function` `between` and `describe function` `case`.
Fix incorrect `between` usage.

yongtang commented:
Thanks @yhuai. The pull request has been updated. Please let me know if there are any other issues.

SparkQA commented:
**[Test build #54928 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54928/consoleFull)** for PR 12128 at commit [`353a132`](https://github.com/apache/spark/commit/353a1320a4ca4d9b1ccbc336d8f5613a21e47a59).

SparkQA commented:
**[Test build #54928 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54928/consoleFull)** for PR 12128 at commit [`353a132`](https://github.com/apache/spark/commit/353a1320a4ca4d9b1ccbc336d8f5613a21e47a59).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54928/
Test FAILed.

hvanhovell commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:
```diff
@@ -213,8 +213,8 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
-      "Usage: ~ n - Bitwise not")
+      "Class: org.apache.spark.sql.catalyst.expressions.BitwiseNot",
+      "Usage: To be added.")
   }
```
Could you also add a test case for one of the hard coded functions?

yongtang replied:
@hvanhovell Thanks. Additional test cases have been added.

yongtang committed:
[SPARK-14335][SQL] Describe function command returns wrong output because some of built-in functions are not in function registry.

Add tests to cover hard coded 'describe function's.

SparkQA commented:
**[Test build #54986 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54986/consoleFull)** for PR 12128 at commit [`927272c`](https://github.com/apache/spark/commit/927272cdc79b7fd8908faea80e8d842d40c2b468).

SparkQA commented:
**[Test build #54986 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54986/consoleFull)** for PR 12128 at commit [`927272c`](https://github.com/apache/spark/commit/927272cdc79b7fd8908faea80e8d842d40c2b468).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54986/
Test PASSed.

yhuai commented:
@yongtang This PR looks good to me. Can you rebase?

SparkQA commented:
**[Test build #55436 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55436/consoleFull)** for PR 12128 at commit [`4c59c81`](https://github.com/apache/spark/commit/4c59c819dabf6cc37248b0c5c1c45a0022e7668c).

SparkQA commented:
**[Test build #55436 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55436/consoleFull)** for PR 12128 at commit [`4c59c81`](https://github.com/apache/spark/commit/4c59c819dabf6cc37248b0c5c1c45a0022e7668c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55436/
Test PASSed.

yongtang commented:
Thanks @yhuai. The pull request has been rebased and passed the Jenkins. Please let me know if there are any other issues.
